### DISCUSSION TITLE
Alternative Attribute Syntax for Command Options and Arguments for better type support and more

### DISCUSSION THREAD

**Discussion by thettler** [laravel/framework#40908] [Ideas]
## The Idea
Using properties in combination with Attributes to declare Arguments and Options on Artisan Commands. 

![image](https://user-images.githubusercontent.com/12132389/153309911-40360b1d-a444-4121-be41-79693f5f3588.png)

## Why? 
The `$signature` method is a cool way to quickly define Arguments and Options for your Artisan commands, but it lacks types and isn't that intuitive. I often have to look in the docs to remember the exact syntax. 

### The benefits of defining Arguments and Options as properties: 
- first of all properties can, and in this case, should be typed. This is great for static analysis and removes a lot of checks down the line.
- The Attributes have clear arguments and so the IDE will tell you what options you have without the need to consult the docs.
- Its more easy to add new Features. For Example the Symfony console supports  "Negatable Options" which are not supported in artisan commands yet, if i am not mistaking. Maybe because the `$signature` string has already quite a lot of features that need to be parsed. Also casting can be more easily implemented like in [this Example](https://github.com/thettler/laravel-command-attribute-syntax#casts) .  
- The Attributes and `$signature` can exist in parallel so no old commands need to be changed 

### The downsides: 
- it adds a little bit more complexity to the Command class
- it is a little bit more to type (but you can still use the `$signature` if you want)

## The Implementation
I already build a little Package that implement the Attribute syntax for Artisan commands. Have a look and try it [here](https://github.com/thettler/laravel-command-attribute-syntax). 
If you are curios and want to know more about the detailed syntax and how it could be working, i have written a detailed documentation where i go more into the details.
https://github.com/thettler/laravel-command-attribute-syntax

The Implementation would be a bit different of course.

## ❤️  Thank you very much for taking the time and reading this. 
What do you think about the Idea and do you have any suggestions or different opinions? 😄 
I would happily create a PR for this Feature ☺️

**Comment by ralphjsmit** [comment]
I think that it would be great if this could be put into the core!

**Reply by Wulfheart** [reply]
I think so too!

**Comment by thettler** [comment]
I took the time and prepared a PR for this. It is currently in draft so feel free to have a look and give me feedback 😄 
 #41131

**Comment by thettler** [comment]
For everybody who likes the idea i have created a Package with all of the Features i mentioned in the PR:

## [Laravel Console Toolkit](https://github.com/thettler/laravel-console-toolkit)

It supports 
| Support | Name                  | Description                                                                                                              |
|:-------:|:----------------------|--------------------------------------------------------------------------------------------------------------------------|
|    ✅    | Laravel Features      | Supports everything laravel can do                                                                                       |
|    ✅    | Attribute Syntax      | Use PHP-Attributes to automatically define your inputs based on types                                                    |
|    ✅    | Casting               | Automatically cast your inputs to Enums, Models, Objects or anything you want                                            |
|    ✅    | Validation            | Use the Laravel Validator to validate the inputs from the console                                                        |
|    ✅    | Auto Ask              | If the user provides an invalid value toolkit will ask again for a valid value without the need to run the command again |
|    ✅    | Negatable Options     | Options can be specified as opposites: --dry or --no-dry                                                                 |
|    ✅    | Option required Value | Options can have required values                                                                                         |


Feel free to try it out and give me Feedback.
